sources/niUI/src/Canvas.cpp

```diff
@@ -622,21 +594,37 @@ struct sCanvasVGPathTesselatedRenderer : public cIUnknownImpl<iVGPathTesselatedR
   //! Called to add the path's polygons.
   virtual void __stdcall AddPathPolygons(iVGPolygonTesselator* apTess, const iVGStyle* apStyle, tBool abStroke) {
     const tVec2fCVec* pVerts = apTess->GetTesselatedVertices();
-    if (pVerts) {
+    if (pVerts && !pVerts->empty()) {
       _canvas->SetColorA(mnAddPathPolygons_VertexColor);
       if (mbAddPathPolygons_TexGen) {
         agg::trans_affine paintTrans = AGGGetTransform(
             mVGTransforms[abStroke?eVGTransform_StrokePaint:eVGTransform_FillPaint].ptr());
-        // paintTrans.invert();
-        for (tVec2fCVec::const_iterator it = pVerts->begin(); it != pVerts->end(); ++it) {
-          const tF32 x = it->x;
-          const tF32 y = it->y;
-          tF32 tx = x;
-          tF32 ty = y;
-          paintTrans.transform(&tx,&ty);
-          tx = tx * mvAddPathPolygons_TexGen_Scale.x + mvAddPathPolygons_TexGen_Translation.x;
-          ty = ty * mvAddPathPolygons_TexGen_Scale.y + mvAddPathPolygons_TexGen_Translation.y;
-          _canvas->VertexPT(Vec3f(x,y,0.0f),Vec2f(tx,ty));
+
+        // fit image into the bounding box
+        sVec2f first = pVerts->GetFirst().mVec2f;
+        tF32 min_x, max_x, min_y, max_y;
+        min_x = max_x = first.x;
+        min_y = max_y = first.y;
+        astl::vector<sVec2f> v_transformed(pVerts->GetSize());
+        niLoop(i, pVerts->GetSize()) {
+          sVec2f v = pVerts->Get(i).mVec2f;
+          paintTrans.transform(&v.x,&v.y);
+          v_transformed[i] = v;
+
+          min_x = Min(min_x, v.x);
+          max_x = Max(max_x, v.x);
+
+          min_y = Min(min_y, v.y);
+          max_y = Max(max_y, v.y);
+        }
+
+        tF32 width = max_x - min_x;
+        tF32 height = max_y - min_y;
+
+        for (const auto& v : v_transformed) {
+          tF32 tx = (v.x - min_x) / width;
+          tF32 ty = (v.y - min_y) / height;
+          _canvas->VertexPT(Vec3f(v.x,v.y,0.0f),Vec2f(tx,ty));
```

Comment:
This does seem to solve a real issue but I dont see `mvAddPathPolygons_TexGen_Scale` & `mvAddPathPolygons_TexGen_Translation` being applied anymore.

Reply:
This part is not finished yet, I was adding those back one by one and trying to use parameters fro m the VGPaint settings

Reply:
things like 
<img width="826" alt="image" src="https://github.com/prenaux/niLang/assets/6252832/c4a97c31-5979-4fd7-871a-f8893cffc4fd">
those APIs didn't have effect on the gradient paint currently

Reply:
It possible that the gradient support is/was only partial. It should at least do basic gradient with basic image transform - that's from memory though it was implemented a few years ago...